cpp/src/Ice/ObjectAdapterFactory.cpp

```diff
@@ -39,59 +37,128 @@ IceInternal::ObjectAdapterFactory::shutdown()
+    }
+}
+
+void
+IceInternal::ObjectAdapterFactory::waitForShutdownAsync(function<void()> completed) noexcept
+{
+    assert(completed); // the caller (Communicator) makes sure completed is callable.
+
+    bool executeCallback = false;
+    {
+        lock_guard lock(_mutex);
+        if (_shutdownCompleted)
+        {
+            executeCallback = true; // execute outside the lock
+        }
+        else
+        {
+            // Start the shutdown completed thread if there is no callback yet.
+            if (_shutdownCompletedCallbacks.empty())
+            {
+                assert(!_shutdownCompletedThread.joinable());
+                _shutdownCompletedThread = std::thread{[self = shared_from_this()]()
+                                                       {
+                                                           self->waitForShutdown();
+                                                           vector<function<void()>> callbacks;
+                                                           {
+                                                               lock_guard lg(self->_mutex);
+                                                               callbacks = std::move(self->_shutdownCompletedCallbacks);
+                                                           }
+
+                                                           for (const auto& callback : callbacks)
+                                                           {
+                                                               callback();
+                                                           }
+                                                       }};
+            }
+
+            _shutdownCompletedCallbacks.push_back(std::move(completed));
+        }
+    }
+
+    if (executeCallback)
+    {
+        completed();
+    }
 }
 
 bool
-IceInternal::ObjectAdapterFactory::isShutdown() const
+IceInternal::ObjectAdapterFactory::isShutdown() const noexcept
 {
     lock_guard lock(_mutex);
-
     return _instance == nullptr;
 }
 
 void
-IceInternal::ObjectAdapterFactory::destroy()
+IceInternal::ObjectAdapterFactory::destroy() noexcept
 {
     //
     // First wait for shutdown to finish.
     //
     waitForShutdown();
 
     list<shared_ptr<ObjectAdapterI>> adapters;
-
+    std::thread shutdownCompletedThread;
     {
         lock_guard lock(_mutex);
+        // We make a copy because we want all callers of destroy to wait for the destruction of these adapters.
         adapters = _adapters;
+
+        shutdownCompletedThread = std::move(_shutdownCompletedThread); // take ownership
+    }
+
+    if (shutdownCompletedThread.joinable())
+    {
+        if (shutdownCompletedThread.get_id() == std::this_thread::get_id())
+        {
+            // destroy is called from a shutdown completed callback - this is fine, but the thread can't join itself.
+            shutdownCompletedThread.detach();
+        }
+        else
+        {
+            shutdownCompletedThread.join();
```

Comment:
Could we just join the tread in the ObjectAdapterFactory destructor?

Reply:
It's nicer to join it in destroy (= earlier), and we would have the same join/detach issue in the destructor: the last refcount can be released by the shutdown completed thread.